toolchain/check/handle_binding_pattern.cpp

```diff
@@ -178,8 +192,16 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
       // formed its initializer.
       // TODO: For general pattern parsing, we'll need to create a block to hold
       // the `let` pattern before we see the initializer.
-      auto bind_id = context.AddPlaceholderInstInNoBlock(
-          make_bind_name(cast_type_id, SemIR::InstId::Invalid));
+      SemIR::LocIdAndInst bind_name =
+          make_bind_name(cast_type_id, SemIR::InstId::Invalid);
+      auto bind_id = context.AddPlaceholderInstInNoBlock(bind_name);
```

Comment:
The `AddInst` call below creates a constant for `BindingPattern` which relies on the constant from this. The reason `AddPlaceholderInstInNoBlock` is updated by `ReplaceInstBeforeConstantUse` is because we can't use it for constants before it's replaced. As a consequence, this code is going to lead to errors.

Also note, the `AddInst` below is going to add the `BindingPattern` to a block in an incorrect order. e.g., look at the ordering of binding_pattern and bind_symbolic_name in toolchain/check/testdata/interface/no_prelude/generic_binding_after_assoc_const.carbon

Rather than trying to add the `BindingPattern` here, I suggest looking at how the binding is added to a block.

Reply:
Let me try to dig a little more into the ordering issue.

Stepping back to the design, what I see is:

```
%x.patt: pattern = binding_pattern `x`, i32
%y.patt: pattern = expr_pattern %y.expected in {
  %GetY: <function> = name_ref @GetY
  %y.expected: i32 = call %GetY()
}
```

So what I'd expect is that `BindingPattern`, under that model, looks like:

```
TypeId type_id;
EntityNameId entity_name_id;
```

Should this be breaking further from the current bind model? e.g., perhaps this should break further from the current binding model?

---

For comparison, the current state is:

```
TypeId type_id;
InstId bind_inst_id;
```

The issue with `bind_inst_id` here is that it refers to a _later_ instruction. So if you think about this in terms of an IR, the execution of this instruction depends on a later instruction.

Note too the ordering isn't always like that. e.g., here's a snippet for `Param`:\

```
// CHECK:STDOUT:     %a.loc11_8.1: i32 = param a
// CHECK:STDOUT:     %.loc11_8: i32 = binding_pattern a, @Add.%a
```

So I think there is inconsistent ordering going on here. I also see you're partly solving this by declaring all binding patterns non-constant. Should we ever be able to constant-evaluate binding patterns? e.g., I'd wonder how that affects something like:

```
fn F(x:! i32) -> bool {
  match (x) {
    case 1 => { return true; }
    default => { return false; }
  }
}
```

In this example, `case 1` is runtime.

Reply:
Note, it's also possible @zygoloid might have separate thoughts here. I'm at least partly trying to pull apart how I'm reading the SemIR there.

Reply:
Sorry, I think I'm somewhat slow at this just because I haven't really thought through the pattern design. Maybe a VC would be helpful to discuss this? I think I have a different understanding of the design, and it may be most helpful to talk through. Would you like to meet this week, or wait for zygoloid to be back since he's more familiar with how patterns work?

Let me ask though, the current design talks about putting `binding_pattern` in one block (a pattern block), and using that to run a match, resulting in `bind_name` in the statement block. In the design's model, `bind_name` is not associated cross-block. What's your plan here? Had you considered starting with that separation, rather than trying to blend it into the current setup and separate later?

There may be a few consequences of using a separate block to consider, for how it would affect this change:

1. The instruction would not be in a statement block, and so should not be seen by lowering.
    - The stub in lowering is only necessary because it's being added to a lowered block.
2. Because this is a separate block, things such as `n` in `(n: i32, n)` would probably want to resolve to the `BindingPattern` rather than a `BindName`.
  - AFAIU this behaves similar to a nested scope, adding lexical name lookup entries that are not accessible later.
3. In the design, there's no clear connection between the `BindingPattern` and a `BindName`.
   - Instead, the pattern is "matched" and results in the `br` code.

Note I'm also realizing looking again at:

```
%x.patt: pattern = binding_pattern `x`, i32
```

I keep getting the structure of `BindingPattern` wrong. To detail how you'd get this SemIR output:

```
// In builtin_inst_kind.def:
CARBON_SEM_IR_BUILTIN_INST_KIND(PatternType, "pattern")

// In typed_insts.h:
struct BindingPattern {
  // The type_id is always BuiltinPatternType.
  TypeId type_id;
  EntityNameId entity_name_id;
  TypeId bind_type_id;
}
```

i.e.:

```
%x.patt: pattern = binding_pattern `x`, i32
                   ^~~~~~~~~~~~~~~ kind == BindingPattern
         ^~~~~~~ type_id == BuiltinPatternType
                                   ^~~ entity_name_id == `x`
                                        ^~~ bind_type_id == i32
```

Is this helpful?

> `binding_pattern` insts aren't ever really executed, though, they're just traversed during the check phase in order to generate the pattern-match IR. `bind_inst_id` isn't an execution dependency of the `binding_pattern`, it's part of the pattern-match IR that we've precomputed and cached.

Are you thinking of execution in terms of lowering to LLVM IR? SemIR may only be re-processed for that today, but for example, we will probably also use SemIR for unused code analysis (stripping out unreferenced instructions) and evaluation of constexpr functions, just as a couple examples. The ability to evaluate SemIR using information accumulation seems important to maintain.

> My intent was to declare that all binding patterns are _non-expressions_. Did I do that wrong?

Where the comment says "These cases are either not expressions or not constant," are you expecting that was written to include patterns? I might just be reading this differently: since the toolchain hasn't implemented actual pattern matching, it simply wasn't considered by the comment; the comment doesn't speak about patterns, positive or negative. Nor does the pattern matching design doc talk about constants.

I'm not sure what the plan for constant evaluation of patterns would be; maybe a TODO? I don't expect it to be resolved by this PR. But, since there's not a design that gets into the deep details, I'm suspicious constants will be needed -- I'd call it an open question.

Reply:
> > Sorry, I think I'm somewhat slow at this just because I haven't really thought through the pattern design. Maybe a VC would be helpful to discuss this? I think I have a different understanding of the design, and it may be most helpful to talk through. Would you like to meet this week, or wait for zygoloid to be back since he's more familiar with how patterns work?
> 
> A VC sounds like a good idea, but my availability for that is uncertain at the moment (I'm WFH covering for a sick co-parent). I'll ping you once I know what my situation is, and in the meantime I'll respond here to what I can.

Sounds good. I'm going to pause review here pending a VC.

Reply:
I just realized the edit you're commenting on doesn't belong in this PR; the `LetIntroducer` case was supposed to be unchanged. I've fixed that now. I don't think that really affects the issues we're discussing in this thread, because they apply just as much to the `TuplePatternStart` case, but please let me know if I've misunderstood that. Otherwise, I'll continue to work on the previously-discussed changes.

Reply:
OK, as discussed elsewhere I've revised this to drop the forward references.